### PR TITLE
UI Developer's Guide

### PR DESCRIPTION
**Pull request by martinpovolny**
Started putting it together. Should have done that sooner.

### CODE REVIEW

File: ui/toolbars.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+instance variables are present and can be used to determine the values of the
+texts.
+
+Furher values have to be passed in as hash. In the example above it is
+  * :url\_parms,
+  * :confirm,
+  * :enabled and
+  * :onwhen.
+
+Very important key is
```
Comment (martinpovolny) [inline ui/toolbars.md]: to show the simple usecase first

File: writing_guides.md [hunk 2]
```diff
@@ -0,0 +1,11 @@
+### Writing this Guide
+
+Docs are written in Mardkown. See e.g. this nice [guide](https://guides.github.com/features/mastering-markdown/).
```
Comment (jrafanie) [inline writing_guides.md]: typo: Markdown

Comment (jrafanie) [inline writing_guides.md]: Prefer "Good examples can be found in this [guide](https://guides.github.com/features/mastering-markdown/)."

File: ui/menus.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
+### Menus
+
+The menu system allow definition of multiple menus. Currently we use 2:
```
Comment (AllenBW) [inline ui/menus.md]: Added s, spelled out two

> The menu system allows for the definition of multiple menus. Currently we use two:

File: ui/menus.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+The main menu styling supports 3 level, that is 2 levels of menu section. Menus are defined in ``app/presenters/menu/default_menu.rb``
+
+#### Menu Sections
+
+```
+Menu::Section.new(:clo, N_("Clouds"), 'fa fa-plus fa-2x', [
+```
+
+Menu sections looks like the example above. It consists of:
```
Comment (AllenBW) [inline ui/menus.md]: Clarity, but maybe tie each of these items to the definitions that follow, either with numbers or maybe **bolding** the leading words?  

> Menu sections, such as the example above consist of the following items: 
> 1.  Identifier
> 2.  Text
> 3.  Icon 
> 4.  Items

File: ui/quadicons.md [hunk 5]
```diff
@@ -0,0 +1,11 @@
+### Quadicons
+
+Quadicons are a condensed graphical way to represent information about various entities and their relations.
+
+Example: 4 VM quadicons:
+
+Quadicons are presen in Explorer trees, in GTLs in Listnavs at least.
```
Comment (AllenBW) [inline ui/quadicons.md]: Added a t 🌮 

> Quadicons are present in, but not limited to, Explorer Trees, GTLs and Listnavs.

File: ui/quadicons.md [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+
+![Textual Summary Example](../images/quadicon.png)
+
+Currently quadicons are rendered using `QuadiconHelper` in `app/helpers/quadicon_helper.rb`.
```
Comment (AllenBW) [inline ui/quadicons.md]: Likely want to uppercase **Quadicons** here.  Its a brand thing so ya might want to take it a step further and bold it everywhere...

File: ui/menus.md [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+Menu::Section.new(:clo, N_("Clouds"), 'fa fa-plus fa-2x', [
+```
+
+Menu sections looks like the example above. It consists of: 
+ * Identifier, 
+ * text, 
+ * icon and 
+ * items.
+
+Identifier is the same as the [RBAC feature](rbac_features.md) that is checked when desicison is make whether the particular section is displayed.
```
Comment (jrafanie) [inline ui/menus.md]: desicison => decision

File: ui/menus.md [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+Text needs to be marked for translation but not translated at the time of definition, therefor ``N_`` see [I18n](i18n.md)
+
+Menu section structure (tree) forms the base of the RBAC tree that is avaliable in Role Editor. (See [RBAC features](rbac_features.md).)
```
Comment (jrafanie) [inline ui/menus.md]: available

File: ui/menus.md [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+Available menu section can be further limited on an appliance via PermissionStores. The Yaml permission store checks the menu section identifier agaist a list of supported RBAC features effectively limiting what is available.
```
Comment (jrafanie) [inline ui/menus.md]: YAML

File: ui/queue_wait.md [hunk 10]
```diff
@@ -0,0 +1,74 @@
+### Running through the Queue and Busy Waiting
+
+General usecase for the UI is to display stuff from the database and queue
+request to the database.
```
Comment (AllenBW) [inline ui/queue_wait.md]: Stuff. _Stuff_ 🍖 🍗 

> General usecase for the UI is to display information from and queue request to the database.

File: ui/menus.md (hunk 9 above)
Comment (jrafanie) [inline ui/menus.md]: agaist => against

File: ui/queue_wait.md [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+The state in the database is maintained by workers as well as various taks that
+are queued are handled by various workers.
+
+The UI calls methods on models or services and and those methods queue things
+that are then done by the workers and later the UI can see the updated state.
+
+Example: If we power off a VM we realy don't do the API call to RHEV or EC2 or
```
Comment (jrafanie) [inline ui/queue_wait.md]: really

File: ui/queue_wait.md [hunk 12]
```diff
@@ -0,0 +1,74 @@
+### Running through the Queue and Busy Waiting
+
+General usecase for the UI is to display stuff from the database and queue
+request to the database.
+
+The state in the database is maintained by workers as well as various taks that
+are queued are handled by various workers.
```
Comment (AllenBW) [inline ui/queue_wait.md]: Transposed word un-transposed

> The state in the database is maintained by workers in addition to various tasks that are queued and handled by various workers.

File: ui/queue_wait.md [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+
+The UI calls methods on models or services and and those methods queue things
+that are then done by the workers and later the UI can see the updated state.
+
+Example: If we power off a VM we realy don't do the API call to RHEV or EC2 or
+VMware straight. But we queue duch action. In the UI we display a flash message
```
Comment (jrafanie) [inline ui/queue_wait.md]: Instead, we queue such an action and display flash message or notification in the UI.

File: ui/queue_wait.md [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+or notification saying that the task was queued.
+
+Later the UI will see the updated state in the database (the VM will be down).
+
+In some cases however it is necessary to wait for the result of such task. The
+task still cannot be done directly as the Appliance running the UI might not be
+even able to reach the API endpoint of the underlying provider.
+
+Examles of this might be:
```
Comment (jrafanie) [inline ui/queue_wait.md]: Examples

File: ui/reports_charts.md [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+- Pct Free Disk
+- Pct Used Disk
+```
+
+Some important attributes:
+ * db: the underlying table,
+ * include: joined tables (meaning JOIN as in SQL),
+ * sortby: columns to use as sort criteria,
+ * cols: columns from the table given in db,
+ * col\_order: order of all solumns,
```
Comment (jrafanie) [inline ui/reports_charts.md]: columns

File: ui/reports_charts.md [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+ * headers: headers of columns (needs to match col\_order),
+ * col\_formats: formating rules for the columns (needs to match col\_order),
```
Comment (jrafanie) [inline ui/reports_charts.md]: formatting

File: ui/queue_wait.md [hunk 17]
```diff
@@ -0,0 +1,74 @@
+### Running through the Queue and Busy Waiting
+
+General usecase for the UI is to display stuff from the database and queue
+request to the database.
+
+The state in the database is maintained by workers as well as various taks that
+are queued are handled by various workers.
+
+The UI calls methods on models or services and and those methods queue things
+that are then done by the workers and later the UI can see the updated state.
```
Comment (AllenBW) [inline ui/queue_wait.md]: Repeated word, clarity

> The UI calls methods on models or services, those methods queue tasks that are then acted on by the workers and result in an updated UI.

File: ui/rbac_features.md [hunk 18]
```diff
@@ -0,0 +1,13 @@
...
+
+RBAC features are defined in `db/fixtures/miq_product_features.yml`. These are
+organized in a tree structure that can be seen in the Role editor under
+Settings.
+
+RBAC features are assigned to roles. For the built-in roles this is predefined in `db/fixtures/miq_user_roles.yml`.
+
+RBAC checking is done in for every action and every button `check_privileges`
+in `ApplicationController`. Therefor each button, each menu item and each
+screen needs to have it's RBAC feature.
```
Comment (AllenBW) [inline ui/rbac_features.md]: Clarity, therefore vs therefor, it's vs its

> RBAC checking is done in for every action and button `check_privileges` in `ApplicationController`. Therefore each button,  menu item and screen needs to have its RBAC feature.

File: ui/reports_charts.md [hunk 19]
```diff
@@ -0,0 +1,97 @@
+### Reports and Charts
+
+Reports are an important part of ManageIQ. Reports feed data not for the 'Reports' section of UI but also:
+ * Charts,
+ * Widgets and
+ * GTL screens.
+
+Reports definitions are stored under `product/reports/`, chars definitions are
+stored under `product/charts/`.
```
Comment (AllenBW) [inline ui/reports_charts.md]: Added missing 🌮 , clarity

> Reports definitions are stored under `product/reports/` and chart definitions under `product/charts/`

File: ui/reports_charts.md [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+
+Reports can be created by hand or through the Report editor in the UI (user reports). In that
+case such report can be dumped and the definition file can be created and
+included with the product.
+
+#### Charts
+
+Data-wise there are 2 types of chars:
+ * charts that visualise count number of discrete values in a report,
+ * charts that visualise exact values from a report.
```
Comment (AllenBW) [inline ui/reports_charts.md]: Spelling of visualise, should be **visualize**  (this row and the previous)

Comment (juliancheal) [inline ui/reports_charts.md]: 🇬🇧

File: ui/reports_charts.md [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+ * sortby: columns to use as sort criteria,
+ * cols: columns from the table given in db,
+ * col\_order: order of all solumns,
+ * headers: headers of columns (needs to match col\_order),
+ * col\_formats: formating rules for the columns (needs to match col\_order),
+ * conditions: selection rules,
+ * group: uses to do GROUP BY.
+
+Reports can be created by hand or through the Report editor in the UI (user reports). In that
+case such report can be dumped and the definition file can be created and
```
Comment (martinpovolny) [inline ui/reports_charts.md]: A way to design a report with a chart might be playing with the editor, then inspecting the result and putting that in the YAML file. Either by dumping it or by inspecting the result in the DB or rails console.

File: writing_guides.md (hunk 2 above)
Comment (martinpovolny) [inline writing_guides.md]: done

File: ui/trees.md [hunk 22]
```diff
@@ -0,0 +1,78 @@
+### Trees
+
+Trees are displayed in various places in the UI. Most notable are trees in explorer-like pages. Other examples include snapshot tree for VMs and RBAC feature tree under Settings/Access Controll/Roles.
```
Comment (martinpovolny) [inline ui/trees.md]: done

File: ui/toolbars.md [hunk 23]
```diff
@@ -0,0 +1,100 @@
+### Toolbars and Buttons
+
+Most screens have toolbars on them. The whole toolbar line is split into up to 3 toolbars.
```
Comment (martinpovolny) [inline ui/toolbars.md]: done

File: ui/toolbars.md [hunk 24]
```diff
@@ -0,0 +1,100 @@
...
+
+There are 3 types of buttons:
+
+ * normal "button",
+ * dropdown "select" and
+ * "twostate".
+
+See `ApplicationHelper::Toolbar::Basic` for details.
+
+The first 4 arguments for any or "button", "select", "twostate" are:
```
Comment (martinpovolny) [inline ui/toolbars.md]: s/or/of

File: ui/toolbars.md [hunk 25]
```diff
@@ -0,0 +1,100 @@
...
+  * :enabled and
+  * :onwhen.
+
+Very important key is
+  * :klass.
+ 
+Toolbars are rendered into JSON by the ManageIQ and then processed by toolbar component from the [ui-components repository](https://github.com/ManageIQ/ui-components)
+
+For some buttons more complex logic is need to decide if those shall be
+displayed, enabled, what are the texts etc. In this case we use the key :klass
```
Comment (martinpovolny) [inline ui/toolbars.md]: done

File: ui/textual_summary.md [hunk 26]
```diff
@@ -0,0 +1,41 @@
+### Textual Summary
+
+Textual summaries display details about a selected entity (such as VM, Host,
+Network, Cluster, Container etc.) and also related entities (Snapshots for VM,
+Hardware for Host etc.).
+
+![Textual Summary Example](../images/textual-summary.png)
+
+Textual summaries consists of groups of related information and each group consists of properties.
```
Comment (martinpovolny) [inline ui/textual_summary.md]: done

File: ui/reports_charts.md (hunk 16 above)
Comment (martinpovolny) [inline ui/reports_charts.md]: done

File: ui/reports_charts.md (hunk 15 above)
Comment (martinpovolny) [inline ui/reports_charts.md]: done

File: ui/reports_charts.md (hunk 20 above)
Comment (martinpovolny) [inline ui/reports_charts.md]: :-D

File: ui/reports_charts.md [hunk 27]
```diff
@@ -0,0 +1,97 @@
...
+
+Data-wise there are 2 types of chars:
+ * charts that visualise count number of discrete values in a report,
+ * charts that visualise exact values from a report.
+
+Charts are created in the Report editor together with reports.
+
+### Dashboard Widgets
+
+Dashboard widgets under cloud inteligence are fed by data from ReportResults that are generated from Reports including charts.
```
Comment (martinpovolny) [inline ui/reports_charts.md]: done

File: ui/menus.md (hunk 8 above)
Comment (martinpovolny) [inline ui/menus.md]: done

File: ui/queue_wait.md (hunk 11 above)
Comment (martinpovolny) [inline ui/queue_wait.md]: done

File: ui/queue_wait.md (hunk 13 above)
Comment (martinpovolny) [inline ui/queue_wait.md]: done

File: ui/queue_wait.md (hunk 14 above)
Comment (martinpovolny) [inline ui/queue_wait.md]: done

File: ui/reports_charts.md (hunk 20 above)
Comment (AllenBW) [inline ui/reports_charts.md]: <3 @juliancheal <3  🐟 🍟

File: README.md [hunk 28]
```diff
@@ -8,9 +8,11 @@
 * [Development Appliance Setup](https://github.com/ManageIQ/manageiq-appliance-dev-setup)
 * [Developer Copr setup for CentOS6](developer_copr_setup_centos6.md)
 * [Internationalization Guidelines](i18n.md)
+* [UI Patterns](ui/patterns.md)
 * [Service UI Skinning](/service_ui/skinning.md)
 * [Running Changelog for Sprints](/community/changelog/)
 * [Roadmap and Sprints](https://github.com/ManageIQ/manageiq/milestones)
+* [Updating this documentation](writing_guides.md)
```
Comment (martinpovolny) [inline README.md]: done

File: ui/calling_api.md [hunk 29]
```diff
@@ -0,0 +1,41 @@
+### Calling the API
+
+Moving forward we want to use the API from the UI, not the other Rails
+controller where possible. To do so when user is logged in besides normal
+session there's also the API token created.
+
+To access the API there's a Angular service available.
+
+To use it you have to inject the "API" into the Angular controller:
+```
```
Comment (martinpovolny) [inline ui/calling_api.md]: done

File: ui/quadicons.md (hunk 5 above)
Comment (martinpovolny) [inline ui/quadicons.md]: done

File: ui/queue_wait.md (hunk 12 above)
Comment (martinpovolny) [inline ui/queue_wait.md]: done

File: ui/queue_wait.md (hunk 17 above)
Comment (martinpovolny) [inline ui/queue_wait.md]: done

File: ui/rbac_features.md (hunk 18 above)
Comment (martinpovolny) [inline ui/rbac_features.md]: done

File: ui/reports_charts.md (hunk 19 above)
Comment (martinpovolny) [inline ui/reports_charts.md]: done

**Comment by martinpovolny** [comment]
Thx for the review. I can add the bold tomorrow or in the text update to this. I am done today.

File: ui/calling_api.md [hunk 30]
```diff
@@ -0,0 +1,46 @@
+### Calling the API
+
+Moving forward we want to use the API from the UI.
+
+To interface with the API, please observe the following best practice:
+
+**Do**: In your Angularcontroller or factory, inject the API service.
```
Comment (himdel) [inline ui/calling_api.md]: > Angularcontroller

Angular controller ;)

File: ui/queue_wait.md [hunk 31]
```diff
@@ -0,0 +1,75 @@
...
+
+In some cases however it is necessary to wait for the result of such task. The
+task still cannot be done directly as the Appliance running the UI might not be
+even able to reach the API endpoint of the underlying provider.
+
+Examples of this might be:
+ * verifying credentials when adding a provider,
+ * opening a remote console to a VM.
+
+In such cases we have the 'wait\_for\_task` call.
```
Comment (himdel) [inline ui/queue_wait.md]: Actually.. this is the problem, opening apostrophe but closing backquote..

File: ui/calling_api.md (hunk 30 above)
Comment (martinpovolny) [inline ui/calling_api.md]: done

File: ui/queue_wait.md [hunk 32]
```diff
@@ -0,0 +1,75 @@
...
+
+Example in `create` we do a call that returns a `task\_id`. Then we call `initiate\_wait\_for\_task`.
+```
```
Comment (martinpovolny) [inline ui/queue_wait.md]: done

File: ui/queue_wait.md (hunk 31 above)
Comment (martinpovolny) [inline ui/queue_wait.md]: done

**Comment by himdel** [comment]
Merged, eager to see the other parts ;)

**Comment by himdel** [comment]
... uh, https://github.com/ManageIQ/guides/pull/165#discussion_r84156431 this one looks like it's still pending..

**Comment by martinpovolny** [comment]
will fix the typo in a next PR

**Comment by hayesr** [comment]
Yay UI documentation! I should mention that in order for these to be accessible on the website we need to update the `guides_menu.yml` file — which I will get around to later this week if no one else does :)

**Pull request merged**